### PR TITLE
Make THashList::RecursiveRemove loop over the list element thread-aware.

### PR DESCRIPTION
**Pull request by pcanal**
Prior to this change, the tranversal might be cut short if one of the RecursiveRemove request the write lock and
thus let another thread modify the list; if this happens some of the node might now have a nullptr payload and
the previous would cut short the traversal.

### CODE REVIEW

**Comment by pcanal** [comment]
@phsft-bot build

File: core/thread/CMakeLists.txt [hunk 1]
```diff
@@ -27,6 +27,8 @@ else()
                          TWin32Thread.cxx TWin32ThreadFactory.cxx)
 endif()
 
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../foundation/res)
```
Comment (amadio) [inline core/thread/CMakeLists.txt]: Hi Philippe, why is this needed? Also, why a relative directory and not `core/foundation/...`?

File: core/thread/inc/ROOT/TReentrantRWLock.hxx [hunk 2]
```diff
@@ -123,6 +124,67 @@ struct RecurseCounts {
...
 
 
+};
+
+struct RecurseCountsLocalSpin {
+   using ReaderColl_t = std::unordered_map<std::thread::id, int>;
+   size_t fWriteRecurse; ///<! Number of re-entry in the lock by the same thread.
+
+   std::thread::id fWriterThread; ///<! Holder of the write lock
+   ReaderColl_t fReadersCount;    ///<! Set of reader thread ids
```
Comment (amadio) [inline core/thread/inc/ROOT/TReentrantRWLock.hxx]: Why do we need this? `std::thread::id` is probably at least in part responsible for slower performance.
All we should care about are how many readers there are, not which threads are reading.

File: core/thread/inc/ROOT/TReentrantRWLock.hxx [hunk 3]
```diff
@@ -123,6 +124,67 @@ struct RecurseCounts {
...
+      GetLocalReadersCount(local) = newvalue;
+   }
+
+   bool IsNotCurrentWriter(local_t &local) { return fWriterThread != local; }
+
+   void SetIsWriter(local_t &local)
+   {
+      // if (fWriteRecurse == std::numeric_limits<decltype(fWriteRecurse)>::max()) {
+      //    ::Fatal("TRWSpinLock::WriteLock", "Too many recursions in TRWSpinLock!");
+      // }
```
Comment (amadio) [inline core/thread/inc/ROOT/TReentrantRWLock.hxx]: No commiting of dead code, please.

File: core/base/src/TDirectory.cxx [hunk 4]
```diff
@@ -550,26 +556,30 @@ void TDirectory::Close(Option_t *option)
...
-   // with a large number of objects (eg >10^5) would take for ever.
-   if (slow) fList->Delete("slow");
-   else      fList->Delete();
+      // Delete objects from directory list, this in turn, recursively closes all
+      // sub-directories (that were allocated on the heap)
+      // if this dir contains subdirs, we must use the slow option for Delete!
+      // we must avoid "slow" as much as possible, in particular Delete("slow")
+      // with a large number of objects (eg >10^5) would take for ever.
+      if (slow) fList->Delete("slow");
+      else      fList->Delete();
```
Comment (amadio) [inline core/base/src/TDirectory.cxx]: This does not follow coding conventions.

File: core/cont/src/THashList.cxx [hunk 5]
```diff
@@ -331,10 +331,22 @@ void THashList::RecursiveRemove(TObject *obj)
...
-   TObject *object;
-   while ((object = next())) {
-      if (object->TestBit(kNotDeleted)) object->RecursiveRemove(obj);
+   // We need to make sure to go through all the node even those
+   // marked as empty by another thread (Eventhough we hold the
+   // read lock if one of the call to RecursiveRemove request
+   // the write lock then the read lock will be suspended and
+   // another thread can modify the list; thanks to the shared_pointer
+   // forward-and-backward links, our view of the list is still intact
+   // but might contains node will nullptr payload)
```
Comment (amadio) [inline core/cont/src/THashList.cxx]: > Eventhough we hold the read lock if one of the call to RecursiveRemove request the write lock then the read lock will be suspended and another thread can modify the list

What? We cannot suspend the read lock. The writer *has* to wait for all readers to be done before writing anything.

File: core/thread/inc/ROOT/TReentrantRWLock.hxx [hunk 6]
```diff
@@ -123,6 +124,67 @@ struct RecurseCounts {
...
+
+   std::thread::id fWriterThread; ///<! Holder of the write lock
+   ReaderColl_t fReadersCount;    ///<! Set of reader thread ids
+
+   std::atomic_flag fSpinLock;    ///<! MSVC doesn't support = ATOMIC_FLAG_INIT;
+
+   using local_t = std::thread::id;
+
+   RecurseCountsLocalSpin() {
+      // MSVC doesn't support fSpinLock=ATOMIC_FLAG_INIT; in the class definition
```
Comment (amadio) [inline core/thread/inc/ROOT/TReentrantRWLock.hxx]: This comment is redundant with the one 5 lines above.

Comment (pcanal) [reply core/thread/inc/ROOT/TReentrantRWLock.hxx]: Sorta.  Again this is copy/paste of similar use and is 'useful' to keep in case somebody chose to outline the function later. (and will not think about re-adding the now not so redundant comment)

File: core/base/src/TDirectory.cxx (hunk 4 above)
Comment (pcanal) [reply core/base/src/TDirectory.cxx]: Existing code ... (and arguably nicer to read that what clang-format says ... but indeed, arguably not as good for debugging).

File: core/cont/src/THashList.cxx (hunk 5 above)
Comment (pcanal) [reply core/cont/src/THashList.cxx]: This is one of the case:

thread1: Take ReadLock; call Stuff; inside Stuff: Try to take the WriteLock
thread2: Take ReadLock; Call Stuff; inside Stuff: Try to take the WriteLock

there we can either have a dead-lock or one of the two threads must yield their read-lock ...

File: core/thread/inc/ROOT/TReentrantRWLock.hxx (hunk 3 above)
Comment (pcanal) [reply core/thread/inc/ROOT/TReentrantRWLock.hxx]: I see this more as a temporarily hidden debugging code :)

**Comment by pcanal** [comment]
@phsft-bot build to see if the success was just luck ...

**Comment by pcanal** [comment]
@phsft-bot build!

File: core/cont/src/THashList.cxx (hunk 5 above)
Comment (amadio) [reply core/cont/src/THashList.cxx]: Yes, readers cannot block after acquiring the lock for any reason. This means that you cannot "call stuff" after acquiring the read lock.

Comment (pcanal) [reply core/cont/src/THashList.cxx]: Yes.  Although adhering strictly to this would mean excessive lock churn, hence the comment/behavior noted above.

**Comment by pcanal** [comment]
@phsft-bot build

**Comment by amadio** [comment]
@phsft-bot build

**Comment by amadio** [comment]
@pcanal The Windows builds now seem to be broken due to these changes:
http://cdash.cern.ch/viewBuildError.php?buildid=429721

**Comment by pcanal** [comment]
@amadio As far as I can tell that is a build dependency problem.  G__Core.cxx should no longer contain the definition of TDirectory::Streamer (i.e. it should have been regenerated but it looks like it did not).

**Comment by amadio** [comment]
The same problem seems to happen on Linux builds:
http://cdash.cern.ch/viewBuildError.php?buildid=429719

**Comment by pcanal** [comment]
@amadio This is consistent with a missing (direct or indirect) dependency of G__Core.cxx on LinkDef1.h in the CMakeList.txt and/or a problem with generating the .d file for G__Core.cxx ...

**Comment by amadio** [comment]
Indeed. Hopefully that will be fixed by #1374.
